### ISSUE TITLE
"Cannot load different versions of Google Charts" doesn't tell me where I attempted to load them, let alone which versions

### ISSUE THREAD

**Issue by php4fan** [google/google-visualization-issues#2835]
This error message:

![image](https://user-images.githubusercontent.com/13969584/86015302-aef1f100-ba21-11ea-9a2f-1209697ee597.png)

shows me 6 levels of backtrace where none of them is in my code.

The **BARE MINIMUM** that I expect from **ANY error message** is to show me the point **in MY CODE** where I did something wrong. The very fact that a backtrace shows a lot of levels inside yoyr library code means you failed to catch an exception that you should have.

So, that would be the bare minimum: show me WHERE I CALLED the load method attempting to load a "different version" of Google Charts.

Then, specifically to this particular error message, since you're telling me that I cannot load different versions, it should tell me **which version** I attempted to load which triggered the error, **AND** which other version had previously been loaded.

Can't believe I have to report this.

**Comment by dlaliberte** [comment]
It's a fair point about reporting the version that you previously loaded and the one you are attempting to load.   
Not many people have ever attempted to load more than one version, so it's not like we get this complaint often.  

Since this is a browser backtrace, we can't necessarily report anything more than the information that JavaScript has.   Why your call of google.charts.load (or google.load, if you are still using jsapi) is missing in the backtrace, I don't know, but I would suspect your call is in a timeout, which effectively removes it from the original call site. You would have to show us enough of your code to help us figure out why.

**Comment by php4fan** [comment]
> Not many people have ever attempted to load more than one version

Nor did I *attempt* to. If I had done that intentionally, I would know where I'm loading each of them (although, I could have forgot, or I could be using code written by different people). If one loads more than one version it's most likely by mistake, which is why it's even more important that the error message be more informative.

Now, it seems that in my particular case the issue is that I'm using both the Charts API and the Analytics Embed API.

I've found this:

https://stackoverflow.com/questions/42858235/conflict-between-google-analytics-embed-api-and-google-charts

Is that hack the best available solution to this day??

**Comment by dlaliberte** [comment]
That StackOverflow thread was 3 years ago.  At the time, the new loader was still very new, and many things have changed since then.   I do have a vague memory of having written that, but forgot the details of the complication with Analytics Embed API.    I don't know if that hack will work, but it is worth a try.

That is the first (and only) case I know of where multiple uses of Google Charts conflict over which version they require.  I think we need a better solution than just throwing an error, even an improved error.  It might be better to just write a warning to the console, and try to continue as if it were the same version.

If you can, you *_should_* update how you load the library to use the new loader.  Full instructions at:    https://developers.google.com/chart/interactive/docs/basic_load_libs#update-library-loader-code

I'm not sure what the Analytics Embed API is doing now, but they should also have updated their loader.     Another big change recently is that now **all** uses of the old jsapi loader redirect to the new loader, and loading version '1' loads the 'current' version.

**Comment by php4fan** [comment]
> If you can, you should update how you load the library to use the new loader. Full instructions at:

I'm already loading the Charts api like that.
What seems to be loaded in an old, obsolete-looking way, is the Analytics API.

And no, unfortunately the Stack Overflow workaround doesn't even work. I have triple-checked I'm doing exactly what the StackOverflow answer says, but I keep getting the same error.

**Comment by dlaliberte** [comment]
Probably the easiest workaround is to figure out which version of Google Charts the Analytics API is loading, and use the same version.  You can find out which version is being loaded by looking in the Network tab in the Chrome debugger (or similar for other browsers).    If they are loading '48.1', it is probably because they are actually loading '1' which is 'current', which is '48.1' at this time.  If so, you would have to load 'current' as well.     Quick internal check, I strongly suspect it is loading '1'.

By the way, the language setting is also a factor in determining which 'version' is loaded.  Just FYI, in case you use it.

**Comment by php4fan** [comment]
> Probably the easiest workaround is to figure out which version of Google Charts the Analytics API is loading

I can't do that. I need the version that I'm loading for the charts that I draw that aren't related to the Analytics API (I don't remember why, but there was a reason I decided to load that version explicitly. It was because of some regression in the chart api that turned my graphs into a mess).

Now here's some more stuff I've figured out that I can't make sense of.

- First https://www.gstatic.com/charts/loader.js gets loaded because of the script tag 
    <script src="https://www.gstatic.com/charts/loader.js"> 
in my html

- then https://www.gstatic.com/charts/45.2/loader.js gets loaded, and from the Network panel of the Developer tools I see that the request loading this js file is initiated by my call to google.charts.load('45.2', ....)

- the exception happens within 45.2/loader.js but the backtrace goes up to /loader.js:

Uncaught (in promise) Error: Cannot load different versions of Google Charts.
    at Object.google.l.m.v.Rl (loader.js:242) // https://www.gstatic.com/charts/45.2/loader.js
    at Object.google.l.m.v.zl (loader.js:244) // https://www.gstatic.com/charts/45.2/loader.js
    at google.l.m.v.load (loader.js:246) // https://www.gstatic.com/charts/45.2/loader.js
    at vc (loader.js:86) // https://www.gstatic.com/charts/loader.js <<<<<<<<< NOTE
    at loader.js:88 // https://www.gstatic.com/charts/loader.js

- This is with the SO code loading the Analytics API (or with the old code without the changed suggested in the SO answer). If I remove all the Analytics-related code, then the two requests above happen in the same order, but there is no error.

- The error prevents the Analytics-related code from drawing the Analytic-related stuff, but it does NOT prevent my non-analytics charts that use the Charts API from working. This suggests that the error is triggered by code in the Analytics API, which not only doesn't trace up to my own code, but not even to Analytics API code causing it.

Is Google Chrome only capable of showing a limited number of items in the backtrace???

**Comment by dlaliberte** [comment]
JavaScript can only show the frames of the backtrace that actually exist.  A timeout creates a break in the chain, thus starting a new chain.  It may be that the use of promises in the loader internally uses timeouts, but if so, I am not sure I can avoid it.

Once the loader.js has loaded a particular version, such as 45.2, then any other calls to google.load or google.charts.load will *start* to use that same version-specific loader (45.2/loader.js for your case), but then throws an error if the version specified is different.

If you are going to try to keep both your code and the analytics code in the same document, rather than, say, putting one or the other in an iframe, then you have to use the same version, currently.   I can make a change to the version-specific loader to not throw an error, but only put out a warning, but you are likely to get some unexpected behavior one way or another.

Maybe we should try to deal with whatever mess your charts experienced, so you can use the 'current' version.

**Comment by php4fan** [comment]
> If you are going to try to keep both your code and the analytics code in the same document, rather than, say, putting one or the other in an iframe, then you have to use the same version, currently

I understand that, but is there a way to tell Analytics to use a specific version of Charts? So I could have Analytics use the same version I'm using, rather than the other way around.

**Comment by dlaliberte** [comment]
No way to change what the code is trying to load.  At best, I can have the loader ignore the version in subsequent load calls, and proceed as if it will work.

I will try to contact the Analytics team and offer to get them to update their loader code, but that will not help you by itself since you could still be trying to load different versions.

Another much larger change is to somehow run Google Charts in a separate sandbox, like an iframe, but display on the main page.   That way, multiple different versions can be loaded in each iframe within the same page.   This is a possibility, but a lot will likely have to be changed to make it work.

**Comment by php4fan** [comment]
> Maybe we should try to deal with whatever mess your charts experienced, so you can use the 'current' version.

Yeah that would definitely be better. It's this one:

https://github.com/google/google-visualization-issues/issues/2673

> JavaScript can only show the frames of the backtrace that actually exist....

Yeah no, well, Chrome does have a limit. I figured out you can tell it to show unlimited traces and here it is, assuming it is now really unlimited. It goes up to the <script> tag dynamically inserted by the Analytics code. If it runs right when the script tag is inserted, that would explain why the backtrace doesn't go further up.

```
Uncaught (in promise) Error: Cannot load different versions of Google Charts.
google.l.m.v.Rl @ loader.js:242
google.l.m.v.zl @ loader.js:244
google.l.m.v.load @ loader.js:246
vc @ loader.js:86
(anonymous) @ loader.js:88
Promise.then (async)
Dc @ loader.js:88
E.load @ loader.js:89
yc @ loader.js:95
d @ cb=gapi.loaded_0:formatted:21856
SE @ cb=gapi.loaded_0:formatted:21856
_.h.xL @ cb=gapi.loaded_0:formatted:22203
_.h.Co @ cb=gapi.loaded_0:formatted:12335
_.oC.xc @ cb=gapi.loaded_0:formatted:18242
(anonymous) @ cb=gapi.loaded_0:formatted:22027
e.Sr @ cb=gapi.loaded_0:formatted:4693
lk @ cb=gapi.loaded_0:formatted:4789
gk @ cb=gapi.loaded_0:formatted:4789
_.Rj.hU @ cb=gapi.loaded_0:formatted:4789
Qj @ cb=gapi.loaded_0:formatted:4524
Promise.then (async)
Lj @ cb=gapi.loaded_0:formatted:4524
_.Pj @ cb=gapi.loaded_0:formatted:4524
hk @ cb=gapi.loaded_0:formatted:4742
_.Rj.th @ cb=gapi.loaded_0:formatted:4731
(anonymous) @ cb=gapi.loaded_0:formatted:4558
d @ cb=gapi.loaded_0:formatted:18111
(anonymous) @ cb=gapi.loaded_0:formatted:16584
(anonymous) @ cb=gapi.loaded_0:formatted:16329
(anonymous) @ cb=gapi.loaded_0:formatted:3837
v.__cb @ cb=gapi.loaded_0:formatted:2973
d @ cb=gapi.loaded_0:formatted:2806
b @ cb=gapi.loaded_0:formatted:2710
```

**Comment by dlaliberte** [comment]
Added a comment to #2673 about the minSpacing option.  

I was thinking of mentioning the Chrome debugger option to show all the backtraces.   I actually try to avoid digging around in backtraces, especially in compiled code.

My idea for changing the version-specific loader to allow different versions (by ignoring the difference) will not work, by the way, since I cannot change the old version-specific loaders.  I'll instead have to do this check in the top-level loader, and change the version to whatever it was last time, so the version-specific loader will never see any difference.

**Comment by php4fan** [comment]
I have changed my Charts-related code to load version `'current'` (while leaving the Analytics-related code untouched), and I keep getting the same error _"Cannot load different versions..."_, now triggered from the 48.1 loader.

Additionally now I have this error, before that:
> Refused to load the script 'https://apis.google.com/js/googleapis.proxy.js?onload=startup' because it violates the following Content Security Policy directive: "script-src 'report-sample' 'nonce-htnFokzOsyQDezENKiwZcg' 'unsafe-inline' 'strict-dynamic' https: http:". 'strict-dynamic' is present, so host-based whitelisting is disabled. Note that 'script-src-elem' was not explicitly set, so 'script-src' is used as a fallback.

Don't know whether it's related.

**Comment by php4fan** [comment]
Apparently I needed to remove this from the Analytics code:
```
<script type="text/javascript" src="https://www.google.com/jsapi"></script>
```

This fixes the "refused to load..." error, but I still get the "Cannot load different versions..." error.

**Comment by dlaliberte** [comment]
Yes, removing the jsapi loader is probably wise anyway, since it results in a redirect to the new gstatic loader.  You don't want to load the loader more than one time, though it tries to survive.   

Since the Analytics code is calling google.load('visualization', '1', ...) and that maps to 'current', then if you are trying to use a different version in your own google.charts.load() call, that will be a conflict, currently.  I am pursuing a change to just use the first version loaded, as I mentioned, but in the meantime, you might try a hack that replaces the google.load function so you can change the version.  Add the following to your code.  It must be evaluated after you have loaded the new loader, and before the Analytics code calls google.load():

```
google.load = function(module, version, settings) {
  google.charts.load('45.2', settings);
}
```

**Comment by php4fan** [comment]
> Since the Analytics code is calling google.load('visualization', '1', ...) and that maps to 'current', then if you are trying to use a different version in your own google.charts.load() call, that will be a conflict, currently.

However, as I mentioned, the issue persist **even if I change my google.charts.load call to use `'current'`**.

> you might try a hack that replaces the google.load function so you can change the version. Add the following to your code. **It must be evaluated after you have loaded the new loader, and before the Analytics code calls google.load()**

Interesting. I wonder if that's somehow similar to the workaround suggested in the Stack Overflow answer cited above. That seemed to do things in the opposite order, I wonder if that's why it didn't work. Just speculating, it might be a completely different hack. 

Right now I'm more interested in updating my charts code to use the current version (i.e. working around #2673 by painfully figuring out every single parameter I need to explicitly set, because out-of-the-box it isn't capable of drawing a small yet legible graph). Actually I expected that to make the version conflict disappear immediately without the need for any hacks.

**Comment by dlaliberte** [comment]
Maybe you need to do your load with google.charts.load('1', ...) so it looks like the same version as what Analytics loads.  That makes sense because the version comparison is done before the mapping (e.g. '1' -> 'current' -> '48.1') rather than after.

**Comment by php4fan** [comment]
> Maybe you need to do your load with google.charts.load('1', ...) 

No, fortunately that doesn't fix it. That would have been embarassing.

**Comment by php4fan** [comment]
If I don't call `google.charts.load()` at all, and let Analytics do its job (that works, obviously then my code that uses Charts doesn't), is there a way I can check what version has actually been loaded? I know it's supposed to be the latest, which appears to be 48.1, but if "supposed to be" and "is" coincided, we wouldn't be here in the first place.

EDIT: never mind, it does look like it's 48.1, since /.../48.1/loader.js is loaded.

**Comment by php4fan** [comment]
Really funny. I have tried replacing my:

```
google.charts.load('1', {'packages': ['corechart', 'line'], 'language': 'es'});
google.charts.setOnLoadCallback(function () {
	// Draw my charts....
});
```
with:

```
google.load('visualization', '1', {'packages': ['corechart', 'line'], 'language': 'es',
						callback: function () {
						//Draw my charts...
						}
					});
```

and my charts still get drawn, and I still get the error "can't load different versions..."

**Comment by php4fan** [comment]
**OH MY F\*\*\*\*\*\* G\*\***

It turns out the culprit was the language:
```
'language': 'es'
```

If I just remove that, everything works. And I can even load whatever version I want - well, I can load 45.2 and it does load that one (haven't checked whether it has some side effects on Analytic's graphs but they generally work).


Now this proves my original point that **if only I had had a decent error message** I wouldn't have wasted two days on this cr\*\*.

In fact, even this: "Can't load different **versions or lanugaues**", which would still be an outrageously vague error message, would have been enough to make me think about the language.

**Comment by php4fan** [comment]
How do I tell Analytics to load a particular language?

**Comment by dlaliberte** [comment]
I don't know how Analytics determines what language to use.  It might use whatever the browser's preferred language is.  Or it might hardcode to something different like 'en'.  

I did mention earlier that the language is also used.    

Very soon, maybe today, this is what you will see:

.. if the version is different:
    console.warn(
        ` Attempting to load version '${version}' of Google Charts,` +
        ` but the previously loaded '${loadedVersion}' will be used instead.`);

.. if the language is different:
    console.warn(
        ` Attempting to load Google Charts for language '${language}',` +
        ` but the previously loaded '${loadedLanguage}' will be used instead.`);

**Comment by php4fan** [comment]
Please also fix Analytics Embed so it doesn't force me to use English in my unrelated charts that happen to live in the same page.

I don't even care about the Analytics graphs being in English ([apparently](https://stackoverflow.com/questions/32436141/google-analytics-embed-api-set-language) Analytics Embed in languages other than English just doesn't exist), I just need my other graphs to have the proper decimal separator for my country.

**Comment by php4fan** [comment]
By the way, if I wanted to have two charts (just charts, no analytics) in the same page in two different languages, wouldn't that be possible??

**Comment by dlaliberte** [comment]
I'll ask Analytics Embed what they can do.  But there is still a general problem with multiple uses of Google Charts in the same page, so, no, you can't do two charts in different languages at this point.  

The only possibility is to somehow isolate the different versions in different iframes, or some equivalent mechanism, while drawing in the container window.  You could also use iframes yourself, such that you load different versions/languages of the charts library in each iframe, and draw in the same iframe.   

Note, there is an option for corecharts which lets you draw the chart by itself in an iframe, but the code still lives in the main page, so that won't help with loading multiple versions/languages in the same page.

**Comment by php4fan** [comment]
> so, no, you can't do two charts in different languages at this point.

This is OT with respect to the issue at hand here, but it's astonishing how a bad design that is.
You should be able to load multiple languages just lile you load multiple packages (either at the same time or in multiple load() calls - by "either" I mean you should be able to do whichever you like, just like with packages), and then _when you draw_ a chart you should be able to choose which language to draw it with.

**Comment by dlaliberte** [comment]
I never liked the current design, which compiles in the localizations for each language in to each package of the library.  The reason it happened that way is that the localizations affect the compiled binaries, shifting the shortened (obfuscated) names of symbols, such that all packages must be compiled as a unit so they can consistently refer to each other.  Extracting localized strings so they can be isolated in a separate language package is part of the issue, which should be relatively easy to handle in a way that avoids this problem.  The more difficult part is how numbers and dates and times are formatted, since formatting is more than just patterns (as strings), but a set of configuration options that is entangled in code.   It's been a long-term goal to fix this, but rarely is there ever a request to display charts in multiple languages on the same page, and the workaround of requiring users to isolate them with iframes is relatively easy, so working on this feature usually gets a lower priority.

**Comment by 1994yashkhandelwal** [comment]
Is anybody can suggest solution for this:  **Uncaught (in promise) Error: Bad version: current**

[Error occurring on chrome extension "Uncaught (in promise) Error: Bad version: current" when using Google Chart JS #2951
](https://github.com/google/google-visualization-issues/issues/2951)

**Comment by FKPSC** [comment]
> **OH MY F****** G****
> 
> It turns out the culprit was the language:
> 
> ```
> 'language': 'es'
> ```
> 
> If I just remove that, everything works. And I can even load whatever version I want - well, I can load 45.2 and it does load that one (haven't checked whether it has some side effects on Analytic's graphs but they generally work).
> 
> Now this proves my original point that **if only I had had a decent error message** I wouldn't have wasted two days on this cr**.
> 
> In fact, even this: "Can't load different **versions or lanugaues**", which would still be an outrageously vague error message, would have been enough to make me think about the language.

My team has had an error in the console bothering us for month:

```
Uncaught TypeError: Cannot set properties of undefined (setting 'FI')
    at jsapi_compiled_i18n_geo_module__de.js:3:21
(anonymous) @ jsapi_compiled_i18n_geo_module__de.js:3
```

It took me about an hour to figure out the issue based the error message. I would still be searching if it wasn't for seeing this. 

What made it harder was that this was happening in a very specific case:
- We have 2 charts on the page, only the second has a language set
- When the first chart was commented out, the issue went away, leading me to believe the issue was with the first chart.
- After playing a game of elimination, I gave up
- I tried searching for the error online with no good luck, until a found this page and decided to try removing the language property on the second chart.
- I stood up and took a walk around the office because I just couldn't anymore....

I wanted to write this down in case it might help anyone. If anyone sees this and it helps, react to the post :)

If anyone working on this sees this, for the love of human sanity, please have a check somewhere and print little more info for this error...
